File: csharp/src/Toolnexus/ContentPart.cs
using System.Text.Json.Serialization;

namespace Toolnexus;

/// <summary>
/// §1B ContentPart — the one non-text channel, shared by every port. A part is
/// <c>text | image | file | audio</c>; a non-text part carries a <c>mimeType</c> plus
/// <b>exactly one</b> of <c>data</c> (standard base64, padded, no line breaks) or <c>url</c>.
///
/// <para><b>A part never holds a filesystem path.</b> A path does not survive a persisted and
/// replayed transcript, nor the MCP / A2A process boundary — so the edge constructors
/// (<see cref="FromFile(string,string?,long?)"/>, <see cref="FromBytes(byte[],string,string?,long?)"/>,
/// <see cref="FromUrl"/>) read and base64 the bytes at construction time.</para>
///
/// <para><b>Accept broadly, store narrowly.</b> The edge takes the objects a .NET caller already
/// holds — a path <c>string</c>, a <see cref="FileInfo"/>, a <see cref="Stream"/>,
/// <c>byte[]</c>/<see cref="ReadOnlySpan{T}"/>/<see cref="ReadOnlyMemory{T}"/>, a <c>data:</c> URL
/// or an <c>https:</c> URL — and what lands in the part is always just a <c>mimeType</c> and
/// base64 <c>data</c>. A stream is consumed eagerly and is <b>not</b> disposed: it stays the
/// caller's to close.</para>
///
/// <para>Wire keys are pinned lower-camel, exactly as <see cref="Request"/> (§10) pins its own —
/// this data crosses languages and processes unchanged.</para>
/// </summary>
public sealed record ContentPart
{
    /// <summary>"text" | "image" | "file" | "audio".</summary>
    [JsonPropertyName("type")]
    public string Type { get; init; } = "";

    /// <summary>Present only on a <c>text</c> part.</summary>
    [JsonPropertyName("text")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Text { get; init; }

    /// <summary>Spelled <c>mimeType</c> in every port and on the wire — never <c>media_type</c>.</summary>
    [JsonPropertyName("mimeType")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? MimeType { get; init; }

    /// <summary>Standard base64 (RFC 4648 §4), padded, no line breaks. Never logged.</summary>
    [JsonPropertyName("data")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Data { get; init; }

    /// <summary>An <c>https:</c> URL. A <c>data:</c> URL is normalised into <c>mimeType</c>+<c>data</c>.</summary>
    [JsonPropertyName("url")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Url { get; init; }

    /// <summary>Optional filename for a <c>file</c> part (the provider's <c>filename</c>).</summary>
    [JsonPropertyName("name")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Name { get; init; }

    // ------------------------------------------------------------------ media table (§6)

    /// <summary>
    /// The fixed media extension table (SPEC §6) — identical in every port. Mime types are
    /// <b>never</b> sniffed from content and <b>never</b> resolved through a platform mime
    /// database (<c>/etc/mime.types</c> varies per machine and would break cross-port parity).
    /// </summary>
    public static readonly IReadOnlyDictionary<string, (string MimeType, string Type)> MediaTable =
        new Dictionary<string, (string, string)>(StringComparer.OrdinalIgnoreCase)
        {
            ["png"] = ("image/png", "image"),
            ["jpg"] = ("image/jpeg", "image"),
            ["jpeg"] = ("image/jpeg", "image"),
            ["gif"] = ("image/gif", "image"),
            ["webp"] = ("image/webp", "image"),
            ["pdf"] = ("application/pdf", "file"),
            ["mp3"] = ("audio/mpeg", "audio"),
            ["wav"] = ("audio/wav", "audio"),
        };

    /// <summary>Look up a path's extension in <see cref="MediaTable"/>; null when unknown.</summary>
    public static (string MimeType, string Type)? MediaFor(string path)
    {
        var ext = System.IO.Path.GetExtension(path ?? "");
        if (ext.Length < 2) return null;
        return MediaTable.TryGetValue(ext[1..], out var hit) ? hit : null;
    }

    /// <summary>The part type a mime type implies: <c>image/*</c>, <c>audio/*</c>, else <c>file</c>.</summary>
    public static string TypeForMime(string mimeType)
    {
        if (mimeType.StartsWith("image/", StringComparison.OrdinalIgnoreCase)) return "image";
        if (mimeType.StartsWith("audio/", StringComparison.OrdinalIgnoreCase)) return "audio";
        return "file";
    }

    // ------------------------------------------------------------------ construction

    /// <summary>
    /// Thrown by the edge constructors for a malformed part: both/neither of
    /// <c>data</c>/<c>url</c>, an unknown extension, or a part over <c>maxPartBytes</c>.
    /// Never carries the part's bytes.
    /// </summary>
    public sealed class InvalidPartException : Exception
    {
        public InvalidPartException(string message) : base(message) { }
    }

    /// <summary>A <c>text</c> part.</summary>
    public static ContentPart FromText(string text) => new() { Type = "text", Text = text };

    /// <summary>Implicit lift so a collection expression may mix strings and parts:
    /// <c>client.RunAsync(["what's in this?", ContentPart.FromFile("shot.png")], toolkit)</c>.</summary>
    public static implicit operator ContentPart(string text) => FromText(text);

    /// <summary>
    /// Read <paramref name="path"/> now and base64 it now — the path never enters the part.
    /// <paramref name="mimeType"/> defaults to the §6 media table; an unknown extension with no
    /// explicit mime type is a typed error naming the extension.
    /// </summary>
    public static ContentPart FromFile(string path, string? mimeType = null, long? maxPartBytes = null)
        => FromBytes(File.ReadAllBytes(path), ResolveMime(path, mimeType),
            System.IO.Path.GetFileName(path), maxPartBytes);

    /// <summary>Async sibling of <see cref="FromFile(string,string?,long?)"/>.</summary>
    public static async Task<ContentPart> FromFileAsync(string path, string? mimeType = null,
        long? maxPartBytes = null, CancellationToken cancellationToken = default)
    {
        var mime = ResolveMime(path, mimeType);
        var bytes = await File.ReadAllBytesAsync(path, cancellationToken).ConfigureAwait(false);
        return FromBytes(bytes, mime, System.IO.Path.GetFileName(path), maxPartBytes);
    }

    /// <summary>
    /// Read <paramref name="file"/> now and base64 it now — the <see cref="FileInfo"/> handle
    /// never enters the part. The mime type defaults to the §6 media table, keyed on
    /// <see cref="FileSystemInfo.Extension"/>.
    /// </summary>
    public static ContentPart FromFile(FileInfo file, string? mimeType = null, long? maxPartBytes = null)
    {
        if (file == null) throw new InvalidPartException("content part: file is required");
        return FromFile(file.FullName, mimeType, maxPartBytes);
    }

    /// <summary>Async sibling of <see cref="FromFile(FileInfo,string?,long?)"/>.</summary>
    public static Task<ContentPart> FromFileAsync(FileInfo file, string? mimeType = null,
        long? maxPartBytes = null, CancellationToken cancellationToken = default)
    {
        if (file == null) throw new InvalidPartException("content part: file is required");
        return FromFileAsync(file.FullName, mimeType, maxPartBytes, cancellationToken);
    }

    /// <summary>
    /// Read <paramref name="stream"/> to the end <b>now</b> and base64 it now — a part never holds
    /// an unread stream, which would not survive a persisted transcript any better than a path.
    ///
    /// <para>Pass <paramref name="mimeType"/> explicitly, or a <paramref name="name"/> whose
    /// extension the §6 media table recognises; neither is a typed error.</para>
    ///
    /// <para><b>The stream is NOT disposed.</b> It belongs to the caller, who is free to keep
    /// reading it or wrap it in their own <c>using</c>. It is read forward from its current
    /// position and never seeked, so a non-seekable stream (a pipe, a network stream) works.</para>
    /// </summary>
    public static ContentPart FromStream(Stream stream, string? mimeType = null, string? name = null,
        long? maxPartBytes = null)
    {
        var mime = ResolveStreamMime(stream, mimeType, name);
        PreflightMax(stream, maxPartBytes);
        using var buffer = new MemoryStream();
        stream.CopyTo(buffer);
        return FromBytes(buffer.GetBuffer().AsSpan(0, (int)buffer.Length), mime, name, maxPartBytes);
    }

    /// <summary>Async sibling of <see cref="FromStream(Stream,string?,string?,long?)"/>. Same
    /// contract: read eagerly, read forward, and never dispose the caller's stream.</summary>
    public static async Task<ContentPart> FromStreamAsync(Stream stream, string? mimeType = null,
        string? name = null, long? maxPartBytes = null, CancellationToken cancellationToken = default)
    {
        var mime = ResolveStreamMime(stream, mimeType, name);
        PreflightMax(stream, maxPartBytes);
        using var buffer = new MemoryStream();
        await stream.CopyToAsync(buffer, cancellationToken).ConfigureAwait(false);
        return FromBytes(buffer.GetBuffer().AsSpan(0, (int)buffer.Length), mime, name, maxPartBytes);
    }

    /// <summary>Base64 native bytes now. <paramref name="mimeType"/> is required.</summary>
    public static ContentPart FromBytes(byte[] bytes, string mimeType, string? name = null,
        long? maxPartBytes = null)
    {
        if (bytes == null) throw new InvalidPartException("content part: bytes are required");
        return FromBytes(bytes.AsSpan(), mimeType, name, maxPartBytes);
    }

    /// <summary>
    /// Span sibling of <see cref="FromBytes(byte[],string,string?,long?)"/>, for a caller who
    /// already holds a slice, a pooled buffer or a stack-allocated span and should not have to
    /// copy it into a fresh array first.
    /// </summary>
    public static ContentPart FromBytes(ReadOnlySpan<byte> bytes, string mimeType, string? name = null,
        long? maxPartBytes = null)
    {
        if (string.IsNullOrEmpty(mimeType)) throw new InvalidPartException("content part: mimeType is required for bytes");
        EnforceMax(bytes.Length, maxPartBytes);
        return new ContentPart
        {
            Type = TypeForMime(mimeType),
            MimeType = mimeType,
            Data = Convert.ToBase64String(bytes),
            Name = name,
        };
    }

    /// <summary>Memory sibling of <see cref="FromBytes(byte[],string,string?,long?)"/>.</summary>
    public static ContentPart FromBytes(ReadOnlyMemory<byte> bytes, string mimeType, string? name = null,
        long? maxPartBytes = null)
        => FromBytes(bytes.Span, mimeType, name, maxPartBytes);

    /// <summary>
    /// A URL part. A <c>data:&lt;mime&gt;;base64,&lt;b64&gt;</c> URL is parsed into
    /// <c>{mimeType, data}</c> at construction, so two spellings of the same bytes never diverge
    /// downstream; an <c>https:</c> URL is kept as <c>url</c>.
    /// </summary>
    public static ContentPart FromUrl(string url, string? mimeType = null, string? name = null,
        long? maxPartBytes = null)
    {
        if (string.IsNullOrEmpty(url)) throw new InvalidPartException("content part: url is required");
        if (url.StartsWith("data:", StringComparison.OrdinalIgnoreCase))
        {
            var comma = url.IndexOf(',');
            if (comma < 0) throw new InvalidPartException("content part: malformed data: URL");
            var header = url[5..comma];
            var isB64 = header.EndsWith(";base64", StringComparison.OrdinalIgnoreCase);
            var mime = (isB64 ? header[..^7] : header).Trim();
            if (mime.Length == 0) mime = mimeType ?? "text/plain";
            var payload = url[(comma + 1)..];
            byte[] bytes;
            try { bytes = isB64 ? Convert.FromBase64String(payload) : System.Text.Encoding.UTF8.GetBytes(Uri.UnescapeDataString(payload)); }
            catch (FormatException) { throw new InvalidPartException("content part: data: URL payload is not valid base64"); }
            return FromBytes(bytes, mime, name, maxPartBytes);
        }
        var explicitMime = mimeType ?? MediaFor(url)?.MimeType
            ?? throw new InvalidPartException(
                $"content part: no mimeType for url \"{Truncate(url)}\" — pass one explicitly");
        return new ContentPart { Type = TypeForMime(explicitMime), MimeType = explicitMime, Url = url, Name = name };
    }

    /// <summary>
    /// Validate the §1B invariant: a non-text part carries a <c>mimeType</c> and exactly one of
    /// <c>data</c>/<c>url</c>. Both, or neither, is a typed error.
    /// </summary>
    public ContentPart Validate()
    {
        if (Type == "text")
        {
            if (Text == null) throw new InvalidPartException("content part: a text part requires text");
            return this;
        }
        if (Type is not ("image" or "file" or "audio"))
            throw new InvalidPartException($"content part: unknown type \"{Type}\"");
        if (string.IsNullOrEmpty(MimeType))
            throw new InvalidPartException($"content part: a {Type} part requires a mimeType");
        var hasData = !string.IsNullOrEmpty(Data);
        var hasUrl = !string.IsNullOrEmpty(Url);
        if (hasData && hasUrl)
            throw new InvalidPartException($"content part: a {Type} part carries both data and url — exactly one is allowed");
        if (!hasData && !hasUrl)
            throw new InvalidPartException($"content part: a {Type} part carries neither data nor url — exactly one is required");
        return this;
    }

    // ------------------------------------------------------------------ observability

    /// <summary>Decoded byte length of <c>data</c> (0 for a url/text part). Base64 is +33%.</summary>
    public long ByteLength => string.IsNullOrEmpty(Data) ? 0 : Data!.Length / 4L * 3 - Padding(Data!);

    /// <summary>
    /// Byte-derived token charge, so a 5 MB image is not free to the compactor. Never derived
    /// from the <c>mimeType</c> string, which would score that image at ~3 tokens and make it
    /// uncompactable.
    /// </summary>
    public int EstimatedTokens => Type == "text"
        ? ((Text?.Length ?? 0) + 3) / 4
        : (int)Math.Max(85, ByteLength / 750);

    /// <summary>
    /// How a part renders in every log line, error message and §9 event:
    /// <c>{type, mimeType, bytes}</c>. The part's <c>data</c> is NEVER logged.
    /// </summary>
    public string Describe() => Type == "text"
        ? $"{{type:text, chars:{Text?.Length ?? 0}}}"
        : $"{{type:{Type}, mimeType:{MimeType}, bytes:{(Data != null ? ByteLength : 0)}}}";

    public override string ToString() => Describe();

    /// <summary>
    /// The canonical §8A one-liner used where a part must be named in user-visible text but never
    /// rendered — an image-only tool result, a placeholder for an unsupported part:
    /// <c>"&lt;type&gt; (&lt;mimeType&gt;, &lt;bytes&gt; bytes)"</c>. <c>bytes</c> is the DECODED
    /// byte count; a part carrying <c>url</c> instead of <c>data</c> renders it as 0.
    /// </summary>
    public string DescribeInText() => $"{Type} ({MimeType}, {ByteLength} bytes)";

    /// <summary>
    /// The canonical §8A placeholder text for a part a client style could not represent:
    /// <c>"[unsupported &lt;type&gt; part (&lt;mimeType&gt;, &lt;bytes&gt; bytes)]"</c>. Same
    /// zero-for-url rule as <see cref="DescribeInText"/>.
    /// </summary>
    public string UnsupportedPlaceholderText() => $"[unsupported {Type} part ({MimeType}, {ByteLength} bytes)]";

    // ------------------------------------------------------------------ internals

    private static int Padding(string b64)
    {
        var n = 0;
        for (var i = b64.Length - 1; i >= 0 && b64[i] == '='; i--) n++;
        return n;
    }

    private static string ResolveMime(string path, string? mimeType)
    {
        if (!string.IsNullOrEmpty(mimeType)) return mimeType!;
        var hit = MediaFor(path);
        if (hit != null) return hit.Value.MimeType;
        var ext = System.IO.Path.GetExtension(path ?? "");
        throw new InvalidPartException(
            $"content part: unknown extension \"{(ext.Length > 0 ? ext : "(none)")}\" — pass an explicit mimeType");
    }

    private static string ResolveStreamMime(Stream stream, string? mimeType, string? name)
    {
        if (stream == null) throw new InvalidPartException("content part: stream is required");
        if (!stream.CanRead) throw new InvalidPartException("content part: stream is not readable");
        if (!string.IsNullOrEmpty(mimeType)) return mimeType!;
        if (!string.IsNullOrEmpty(name))
        {
            var hit = MediaFor(name!);
            if (hit != null) return hit.Value.MimeType;
            var ext = System.IO.Path.GetExtension(name!);
            throw new InvalidPartException(
                $"content part: unknown extension \"{(ext.Length > 0 ? ext : "(none)")}\" — pass an explicit mimeType");
        }
        throw new InvalidPartException(
            "content part: mimeType is required for a stream — pass one, or a name whose extension the media table knows");
    }

    /// <summary>Fast-fail on an over-size stream we can measure without reading it. The §1B
    /// guarantee stays the assembly check; this is only the better message.</summary>
    private static void PreflightMax(Stream stream, long? maxPartBytes)
    {
        if (maxPartBytes is not > 0 || !stream.CanSeek) return;
        var remaining = stream.Length - stream.Position;
        if (remaining > maxPartBytes.Value)
            throw new InvalidPartException(
                $"content part: {remaining} decoded bytes exceeds maxPartBytes {maxPartBytes.Value}");
    }

    private static void EnforceMax(long bytes, long? maxPartBytes)
    {
        if (maxPartBytes is > 0 && bytes > maxPartBytes.Value)
            throw new InvalidPartException(
                $"content part: {bytes} decoded bytes exceeds maxPartBytes {maxPartBytes.Value}");
    }

    private static string Truncate(string s) => s.Length <= 64 ? s : s[..64] + "…";
}

File: java/src/main/java/io/github/muthuishere/toolnexus/ContentPart.java
package io.github.muthuishere.toolnexus;

import com.fasterxml.jackson.annotation.JsonInclude;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * §1B — one non-text content part: {@code text | image | file | audio}.
 *
 * <p>A flat record with a {@code type} discriminator, mirroring {@link Request} (the port's other
 * wire union) rather than a sealed hierarchy — the wire keys are pinned
 * ({@code type, text, mimeType, data, url, name}) and round-trip through the shared Jackson
 * mapper with no custom serializer.
 *
 * <ul>
 *   <li>a non-text part carries {@code mimeType} plus <b>exactly one</b> of {@code data}
 *       (standard base64, padded, no line breaks) or {@code url} — both, or neither, is a
 *       construction error;</li>
 *   <li><b>a part never holds a filesystem path.</b> A path does not survive a persisted and
 *       replayed transcript, nor the MCP / A2A process boundary. The edge constructors
 *       ({@link #ofFile}, {@link #ofBytes}, {@link #ofUrl}, {@link #ofStream}) normalise at
 *       construction so the path, the stream or the raw bytes never enter the part. They accept
 *       broadly — a {@link Path}, a {@link java.io.File}, an {@link InputStream}, bytes, a
 *       {@code data:} URL — and store narrowly: {@code mimeType} + base64 {@code data}.</li>
 * </ul>
 *
 * <p>The mime type of a path comes from the <b>fixed</b> extension table below (§6 {@code read}),
 * never from magic bytes and never from a platform mime database — {@code /etc/mime.types} varies
 * per machine and would break cross-port parity.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ContentPart(String type, String text, String mimeType, String data,
                          String url, String name) {

    /** The four part types (§1B). */
    public static final String TEXT = "text";
    public static final String IMAGE = "image";
    public static final String FILE = "file";
    public static final String AUDIO = "audio";

    /**
     * The fixed media extension table (§6 {@code read} / §1B edge constructors) — shared with
     * every port. {@code ext -> {mimeType, partType}}. No sniffing, no platform mime database.
     */
    private static final Map<String, String[]> MEDIA = Map.of(
            "png", new String[]{"image/png", IMAGE},
            "jpg", new String[]{"image/jpeg", IMAGE},
            "jpeg", new String[]{"image/jpeg", IMAGE},
            "gif", new String[]{"image/gif", IMAGE},
            "webp", new String[]{"image/webp", IMAGE},
            "pdf", new String[]{"application/pdf", FILE},
            "mp3", new String[]{"audio/mpeg", AUDIO},
            "wav", new String[]{"audio/wav", AUDIO});

    /**
     * Edge limit in <b>decoded</b> bytes (not the base64 string, which is +33%), enforced by the
     * edge constructors. {@code 0} (the default) means unlimited. This is a process-wide setting
     * because the edge constructors are static; {@link LlmClient.Options#maxPartBytes} enforces
     * the same limit again at request assembly, for parts that never passed an edge constructor
     * (an MCP server's, say).
     */
    private static volatile long maxPartBytes = 0;

    /** Set the process-wide edge limit in decoded bytes; {@code 0} = unlimited. */
    public static void setMaxPartBytes(long bytes) {
        maxPartBytes = Math.max(0, bytes);
    }

    /** The current process-wide edge limit in decoded bytes; {@code 0} = unlimited. */
    public static long maxPartBytes() {
        return maxPartBytes;
    }

    /** Canonical, validating construction. Prefer the named factories. */
    public ContentPart {
        if (type == null || type.isEmpty()) {
            throw new IllegalArgumentException("ContentPart: type is required");
        }
        if (TEXT.equals(type)) {
            if (data != null || url != null) {
                throw new IllegalArgumentException("ContentPart: a text part carries neither data nor url");
            }
        } else {
            if (!IMAGE.equals(type) && !FILE.equals(type) && !AUDIO.equals(type)) {
                throw new IllegalArgumentException("ContentPart: unknown part type \"" + type
                        + "\" (expected text, image, file or audio)");
            }
            if (mimeType == null || mimeType.isEmpty()) {
                throw new IllegalArgumentException("ContentPart: a " + type + " part requires a mimeType");
            }
            boolean hasData = data != null && !data.isEmpty();
            boolean hasUrl = url != null && !url.isEmpty();
            if (hasData && hasUrl) {
                throw new IllegalArgumentException(
                        "ContentPart: a " + type + " part carries exactly one of data or url, not both");
            }
            if (!hasData && !hasUrl) {
                throw new IllegalArgumentException(
                        "ContentPart: a " + type + " part carries exactly one of data or url, got neither");
            }
        }
    }

    // ------------------------------------------------------------------
    // Edge constructors — a path / bytes / a data: URL / an https: URL all
    // normalise into a path-free, bytes-or-url part.
    // ------------------------------------------------------------------

    /** A text part. */
    public static ContentPart text(String text) {
        return new ContentPart(TEXT, text == null ? "" : text, null, null, null, null);
    }

    /**
     * Read {@code path} now, base64 it now; the mime type and part type come from the fixed
     * extension table. The part carries no path.
     *
     * <p>Throws {@link UncheckedIOException} on an unreadable file — forcing a try/catch around a
     * literal argument is hostile, and {@code Files.lines}/{@code Files.walk} set the precedent.
     * Use {@link #ofFileChecked} when a checked exception is wanted. An extension outside the
     * table (with no explicit mime type) is an {@link IllegalArgumentException} naming it — never
     * a silent {@code application/octet-stream}.
     */
    public static ContentPart ofFile(Path path) {
        return ofFile(path, null);
    }

    /** {@link #ofFile(Path)} with an explicit mime type, for an extension outside the table. */
    public static ContentPart ofFile(Path path, String mimeType) {
        try {
            return ofFileChecked(path, mimeType);
        } catch (IOException e) {
            throw new UncheckedIOException("ContentPart.ofFile: " + path, e);
        }
    }

    /** {@link #ofFile(Path)} for callers that want the checked {@link IOException}. */
    public static ContentPart ofFileChecked(Path path) throws IOException {
        return ofFileChecked(path, null);
    }

    /** {@link #ofFile(Path, String)} for callers that want the checked {@link IOException}. */
    public static ContentPart ofFileChecked(Path path, String mimeType) throws IOException {
        if (path == null) throw new IllegalArgumentException("ContentPart.ofFile: path is required");
        String mime = mimeType;
        String type;
        if (mime == null || mime.isEmpty()) {
            String[] hit = mediaFor(path.getFileName() == null ? path.toString() : path.getFileName().toString());
            if (hit == null) {
                throw new IllegalArgumentException("ContentPart.ofFile: unknown extension \""
                        + extensionOf(path.toString()) + "\" for " + path
                        + " — pass an explicit mimeType (mime is never sniffed from content)");
            }
            mime = hit[0];
            type = hit[1];
        } else {
            type = typeForMime(mime);
        }
        return ofBytes(type, mime, Files.readAllBytes(path));
    }

    // ------------------------------------------------------------------
    // The file and stream objects a Java caller already holds (§1B: "a port
    // accepts the file and byte objects its users already hold"). Accept
    // broadly, store narrowly — whatever comes in, the part holds only
    // `mimeType` + base64 `data`, never a File, a stream or a path.
    // ------------------------------------------------------------------

    /**
     * {@link #ofFile(Path)} for the {@link java.io.File} most Java callers actually hold —
     * delegates to the {@code Path} logic via {@link java.io.File#toPath()}.
     */
    public static ContentPart ofFile(java.io.File file) {
        return ofFile(file, null);
    }

    /** {@link #ofFile(java.io.File)} with an explicit mime type, for an extension outside the table. */
    public static ContentPart ofFile(java.io.File file, String mimeType) {
        if (file == null) throw new IllegalArgumentException("ContentPart.ofFile: file is required");
        return ofFile(file.toPath(), mimeType);
    }

    /** {@link #ofFile(java.io.File)} for callers that want the checked {@link IOException}. */
    public static ContentPart ofFileChecked(java.io.File file) throws IOException {
        return ofFileChecked(file, null);
    }

    /** {@link #ofFile(java.io.File, String)} for callers that want the checked {@link IOException}. */
    public static ContentPart ofFileChecked(java.io.File file, String mimeType) throws IOException {
        if (file == null) throw new IllegalArgumentException("ContentPart.ofFile: file is required");
        return ofFileChecked(file.toPath(), mimeType);
    }

    /**
     * Read {@code in} <b>fully, now</b>, and base64 it now; the part carries the encoding, never
     * the stream. A part must never hold an unread stream — a half-read stream would not survive
     * a persisted transcript any better than a path does.
     *
     * <p><b>The stream is not closed.</b> It is the caller's; toolnexus reads it and leaves
     * closing to whoever opened it (so a {@code try}-with-resources around the call still works,
     * and a shared stream is not yanked out from under its owner).
     *
     * <p>Throws {@link UncheckedIOException} on a read failure, matching {@link #ofFile(Path)};
     * use {@link #ofStreamChecked} for the checked {@link IOException}.
     */
    public static ContentPart ofStream(InputStream in, String mimeType) {
        try {
            return ofStreamChecked(in, mimeType);
        } catch (IOException e) {
            throw new UncheckedIOException("ContentPart.ofStream", e);
        }
    }

    /**
     * {@link #ofStream(InputStream, String)} for a stream that has a filename: when
     * {@code mimeType} is {@code null} or empty, the mime and part type come from the fixed
     * extension table (§6 {@code read}), exactly as they do for a path. An extension outside the
     * table with no explicit mime is an {@link IllegalArgumentException} naming it. {@code name}
     * also becomes a {@code file} part's display name.
     */
    public static ContentPart ofStream(InputStream in, String name, String mimeType) {
        try {
            return ofStreamChecked(in, name, mimeType);
        } catch (IOException e) {
            throw new UncheckedIOException("ContentPart.ofStream: " + name, e);
        }
    }

    /** {@link #ofStream(InputStream, String)} for callers that want the checked {@link IOException}. */
    public static ContentPart ofStreamChecked(InputStream in, String mimeType) throws IOException {
        if (in == null) throw new IllegalArgumentException("ContentPart.ofStream: stream is required");
        if (mimeType == null || mimeType.isEmpty()) {
            throw new IllegalArgumentException("ContentPart.ofStream: a stream has no extension to "
                    + "read a mime type from — pass an explicit mimeType, or a name "
                    + "(mime is never sniffed from content)");
        }
        return ofBytes(typeForMime(mimeType), mimeType, in.readAllBytes());
    }

    /** {@link #ofStream(InputStream, String, String)} for callers that want the checked {@link IOException}. */
    public static ContentPart ofStreamChecked(InputStream in, String name, String mimeType) throws IOException {
        if (in == null) throw new IllegalArgumentException("ContentPart.ofStream: stream is required");
        String mime = mimeType;
        String type;
        if (mime == null || mime.isEmpty()) {
            String[] hit = mediaFor(name);
            if (hit == null) {
                throw new IllegalArgumentException("ContentPart.ofStream: unknown extension \""
                        + extensionOf(name) + "\" for " + name
                        + " — pass an explicit mimeType (mime is never sniffed from content)");
            }
            mime = hit[0];
            type = hit[1];
        } else {
            type = typeForMime(mime);
        }
        ContentPart p = ofBytes(type, mime, in.readAllBytes());
        return FILE.equals(type) && name != null && !name.isEmpty() ? p.withName(name) : p;
    }

    /** Base64 {@code bytes} now; the part carries the encoding, never the array. */
    public static ContentPart ofBytes(String type, String mimeType, byte[] bytes) {
        if (bytes == null) throw new IllegalArgumentException("ContentPart.ofBytes: bytes are required");
        checkSize(bytes.length);
        return new ContentPart(type, null, mimeType, Base64.getEncoder().encodeToString(bytes), null, null);
    }

    /** An {@code image} part from raw bytes. */
    public static ContentPart image(String mimeType, byte[] bytes) {
        return ofBytes(IMAGE, mimeType, bytes);
    }

    /** An {@code audio} part from raw bytes. */
    public static ContentPart audio(String mimeType, byte[] bytes) {
        return ofBytes(AUDIO, mimeType, bytes);
    }

    /** A {@code file} part from raw bytes, with an optional display name. */
    public static ContentPart file(String mimeType, byte[] bytes, String name) {
        ContentPart p = ofBytes(FILE, mimeType, bytes);
        return name == null || name.isEmpty() ? p : p.withName(name);
    }

    /** A part from an already-base64'd payload (an MCP server's, say) — taken verbatim. */
    public static ContentPart ofBase64(String type, String mimeType, String base64) {
        if (base64 != null) checkSize(decodedLength(base64));
        return new ContentPart(type, null, mimeType, base64, null, null);
    }

    /**
     * A {@code data:<mime>;base64,<b64>} URL is parsed into {@code {mimeType, data}} at
     * construction — never stored as a {@code url}, so two spellings of the same bytes cannot
     * diverge downstream. Any other URL (an {@code https:} one) is kept as {@code url}, and then
     * {@code mimeType} is required.
     */
    public static ContentPart ofUrl(String type, String url) {
        return ofUrl(type, null, url);
    }

    /** {@link #ofUrl(String, String)} with an explicit mime type (required for a non-data URL). */
    public static ContentPart ofUrl(String type, String mimeType, String url) {
        if (url == null || url.isEmpty()) {
            throw new IllegalArgumentException("ContentPart.ofUrl: url is required");
        }
        if (url.startsWith("data:")) {
            int comma = url.indexOf(',');
            int semi = url.indexOf(';');
            if (comma < 0 || semi < 0 || semi > comma
                    || !url.regionMatches(true, semi, ";base64,", 0, ";base64,".length())) {
                throw new IllegalArgumentException(
                        "ContentPart.ofUrl: only data:<mime>;base64,<b64> data URLs are supported");
            }
            String mime = url.substring("data:".length(), semi);
            String b64 = url.substring(comma + 1);
            checkSize(decodedLength(b64));
            return new ContentPart(type, null, mime, b64, null, null);
        }
        return new ContentPart(type, null, mimeType, null, url, null);
    }

    /** A copy carrying {@code name} (the {@code file} part's display name). */
    public ContentPart withName(String newName) {
        return new ContentPart(type, text, mimeType, data, url, newName);
    }

    // ------------------------------------------------------------------
    // Rendering for logs / §9 events — {type, mimeType, bytes}. `data` NEVER.
    // ------------------------------------------------------------------

    /** Decoded byte length of {@code data}, or {@code 0} for a url/text part. */
    public long bytes() {
        return data == null ? 0 : decodedLength(data);
    }

    /** How a part appears in a log line or a §9 event: {@code {type, mimeType, bytes}}, never {@code data}. */
    public Map<String, Object> describe() {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("type", type);
        m.put("mimeType", mimeType);
        m.put("bytes", bytes());
        return m;
    }

    /** {@link #describe()} over a list; {@code null} when there is nothing to describe. */
    public static List<Map<String, Object>> describeAll(List<ContentPart> parts) {
        if (parts == null || parts.isEmpty()) return null;
        List<Map<String, Object>> out = new java.util.ArrayList<>(parts.size());
        for (ContentPart p : parts) out.add(p.describe());
        return out;
    }

    /** Never prints {@code data} — part bytes are user content and stay out of every log line. */
    @Override
    public String toString() {
        return "ContentPart{type=" + type + ", mimeType=" + mimeType + ", bytes=" + bytes()
                + (url != null ? ", url=" + url : "") + "}";
    }

    // ------------------------------------------------------------------
    // Media table helpers (shared with the `read` builtin, §6)
    // ------------------------------------------------------------------

    /** {@code {mimeType, partType}} for a filename's extension, or {@code null} when unlisted. */
    static String[] mediaFor(String filename) {
        String ext = extensionOf(filename);
        return ext.isEmpty() ? null : MEDIA.get(ext);
    }

    static String extensionOf(String filename) {
        if (filename == null) return "";
        int dot = filename.lastIndexOf('.');
        if (dot < 0 || dot == filename.length() - 1) return "";
        return filename.substring(dot + 1).toLowerCase(Locale.ROOT);
    }

    /** The part type an explicit mime type implies: {@code image/*}, {@code audio/*}, else file. */
    static String typeForMime(String mimeType) {
        String m = mimeType == null ? "" : mimeType.toLowerCase(Locale.ROOT);
        if (m.startsWith("image/")) return IMAGE;
        if (m.startsWith("audio/")) return AUDIO;
        return FILE;
    }

    /** Decoded length of a standard base64 string, without decoding it. */
    static long decodedLength(String base64) {
        if (base64 == null || base64.isEmpty()) return 0;
        int len = base64.length();
        int pad = 0;
        if (base64.charAt(len - 1) == '=') pad++;
        if (len > 1 && base64.charAt(len - 2) == '=') pad++;
        return (long) len * 3 / 4 - pad;
    }

    private static void checkSize(long decodedBytes) {
        long limit = maxPartBytes;
        if (limit > 0 && decodedBytes > limit) {
            throw new IllegalArgumentException("ContentPart: part is " + decodedBytes
                    + " decoded bytes, over the maxPartBytes limit of " + limit);
        }
    }
}
